File: sweetfactory-compiler/src/main/java/sweetfactory/compiler/FactoryImplementation.java
package sweetfactory.compiler;

import com.squareup.javapoet.TypeSpec;

public class FactoryImplementation {

    private String packageName;
    private TypeSpec.Builder typeSpecBuilder;

    public FactoryImplementation(String packageName, TypeSpec.Builder typeSpecBuilder) {
        this.packageName = packageName;
        this.typeSpecBuilder = typeSpecBuilder;
    }

    public String getPackageName() {
        return packageName;
    }

    public TypeSpec.Builder getTypeSpecBuilder() {
        return typeSpecBuilder;
    }
}

File: sweetfactory-compiler/src/main/java/sweetfactory/compiler/FactoryImplementation.kt
package sweetfactory.compiler

import com.squareup.javapoet.ClassName
import com.squareup.javapoet.TypeSpec
import javax.lang.model.element.Modifier

internal data class FactoryImplementation(
    val packageName: String,
    val typeSpecBuilder: TypeSpec.Builder
) {
    companion object {
        fun from(fullClassName: String): FactoryImplementation {
            val className = fullClassName.substring(fullClassName.lastIndexOf(".") + 1)
            val packageName = fullClassName.replace(".$className", "")

            val interfaceClass = ClassName.get(packageName, className)

            val builder = TypeSpec.classBuilder(className + "Impl")
                .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                .addSuperinterface(interfaceClass)

            return FactoryImplementation(packageName, builder)
        }
    }
}